### DISCUSSION TITLE
Asynchronous form post on custom Authenticator, how to?

### DISCUSSION THREAD

**Discussion by oculos** [keycloak/keycloak#23291] [Q&A] [answered]
Hi,

I have implemented a custom authenticator to handle 2FA on our external provider.
My Authenticator is based on an HTTPRequest to an internal server to authorize the user.

It's code is pretty simple: 

```
protected boolean validateAnswer(AuthenticationFlowContext context)  {
        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
        MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
        String secret = formData.getFirst("secret_answer");
        //String credentialId = formData.getFirst("credentialId");
        String username = context.getUser().getUsername();


        HttpClient client = HttpClient.newHttpClient();
        String serviceUrl = config.getConfig().get("http-radius.proxy");
        HttpRequest request = HttpRequest.newBuilder()
  .uri(URI.create(serviceUrl))
  .POST(HttpRequest.BodyPublishers.ofString("{\"token\":\""+secret+"\", \"user\":\""+username+"\"}"))
  .build();
          try {
          java.net.http.HttpResponse<String> response = client.send(request, java.net.http.HttpResponse.BodyHandlers.ofString());
          if (response.statusCode() == 200)return true;
         }catch (Exception e){ return false;}

         return false;
}
```

This works great when the user has a numeric token. But in some situations, the user has a push notification. So I changed the form in these cases so that a javascript automatically submits the form so that the above method is called. 

My problem is that when this happens, the HTTP request sent by the browser (ie., the form submission) is hanging expecting the action method to receive a response from the 2FA server. 

I tried use a javascript to post the form asynchronously, but what should I do with the asynchronous response, as nothing happens when something is returned.

Is there something I could do after this: 

```
 fetch(form.action, {
            method: form.method,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            body: params
        })
```

So that the page refreshes according to the new context when it receives the answer of the form that was sent? 

I tried to see what @thomasdarimont did[ here](https://github.com/thomasdarimont/keycloak-extension-playground/tree/master/auth-identity-first-extension), but he modifies the DOM when he receives a new result - I'm not sure that's what I want. I just need to be redirect to the application or to the authentication flow again.

Any clue on what I should do?

**Comment by jonkoops** [comment]
If you want to send form data you need to pass a [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) object as the `body` in your fetch call. For more information see [Using the Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#body).

**Reply by oculos** [reply]
Thanks @jonkoops . I know that (yes, the code up there is incomplete). But my question is actually on how to update the DOM when a reply is received from the `action` method of the Authenticator. Or, with other words, how to actually load the url that was fetched after the form was sent.

**Reply by jonkoops** [reply]
Ah my bad, I don't have a clear answer for you there.

**Reply by oculos** [reply, marked as answer]
It works now. I just add a `then(res =>  window.location = res.url);` to the fetch method, and boom, I got exactly what I need.

Best,
Francis
